[Doc Assembly Branch] Vault 1.9 release (#12944)
hashicorp/vault: website/content/docs/concepts/oidc-provider.mdx

This feature allows clients speaking to the OIDC protocol to take advantage of Vault's various authentication methods and source of identity. Clients can configure their authentication logic to talk to Vault. Once enabled, Vault will act as the bridge to identity providers via its existing authentication methods. Clients will also obtain identity information for their end-users by leveraging custom templating of Vault identity information.

Vault as an OIDC provider allows mutual Vault and Boundary customers to leverage Vault's identity system to delegate authentication and authorization to Vault.  Vault, therefore, acts as an identity provider for Boundary. Other HashiCorp products such as Consul can also leverage Vault's identity system and provide delegated authentication and authorization to its users. Having Vault as an OIDC provider allows a single sign-on experience to their end-users for organizations that want to leverage Vault as an identity provider.

## Configuration Options

The next few sections of the document provide implementation details for each resource that permits Vault configuration as an OIDC identity provider.

### Providers

> A Vault namespace may contain several provider resources. Each configured provider will publish the APIs listed within the OIDC flow. The APIs will be served via backend path-based routing on Vault's listen to address. The APIs will be served via the backend path-based routing on Vault's listen to the [address](/docs/configuration/listener/tcp#address).

Reviewer: Suggest to use the wording `Vault's listen address`

Reviewer: ```suggestion
A Vault namespace may contain several provider resources. Each configured provider will publish the APIs listed within the OIDC flow. The APIs will be served via the backend path-based routing on Vault's listen [address](/docs/configuration/listener/tcp#address).
```

### Scopes

Providers may reference scope resources via the `scopes_supported` parameter to make specific identity information available as claims.

A scope will have the following configuration parameters:

* **Description**: identity information captured by the scope
* **Template**: maps individual claims to Vault identity information

The template parameter takes advantage of the [JSON-based templating](/api-docs/secret/identity/tokens#template) used by identity tokens for claims mapping. This means the parameter will take a JSON string of arbitrary structure where the values may be replaced with specific identity information. Template parameters that are not present for a Vault identity are omitted from the resulting claims without an error.

Example of a JSON template for a scope:

```
{
  "groups": {{identity.entity.groups.names}},
  "email": {{identity.entity.metadata.user_email}},
  "auth0": {

>       "name": {{identity.entity.aliases.$MNT_ACC.metadata.user}}

Reviewer: I am wondering if it makes sense to change `$MNT_ACC` to `$MOUNT_ACCESSOR` for more clarity?

Author: I missed this one initially. Yes, I think changing it to _ACCESSOR will provide that clarity. At first glance, I thought it might've been referring to "Account". I'll make that change.

  }
}
```

The full list of template parameters include:

* `identity.entity.id`
* `identity.entity.name`
* `identity.entity.groups.ids`
* `identity.entity.groups.names`
* `identity.entity.metadata`
* `identity.entity.metadata.<metadata key>`
* `identity.entity.aliases.<mount accessor>.id`
* `identity.entity.aliases.<mount accessor>.name`
* `identity.entity.aliases.<mount accessor>.metadata`
* `identity.entity.aliases.<mount accessor>.metadata.<metadata key>`
* `time.now`
* `time.now.plus.<duration>`

> * `time.now.minus.<duration>`

Reviewer: Maybe we should make this a table with the descriptions similar to https://www.vaultproject.io/docs/secrets/identity#token-contents-and-templates

Reviewer: ```suggestion
| Name                                                                             | Description                                                                             |
| :------------------------------------------------------------------------------- | :-------------------------------------------------------------------------------------- |
| `identity.entity.id`                                                             | The entity's ID                                                                         |
| `identity.entity.name`                                                           | The entity's name                                                                       |
| `identity.entity.groups.ids`                                                     | The IDs of the groups the entity is a member of                                         |
| `identity.entity.groups.names`                                                   | The names of the groups the entity is a member of                                       |
| `identity.entity.metadata`                                                       | Metadata associated with the entity                                                     |
| `identity.entity.metadata.<metadata key>`                                        | Metadata associated with the entity for the given key                                   |
| `identity.entity.aliases.<mount accessor>.id`                                    | Entity alias ID for the given mount                                                     |
| `identity.entity.aliases.<mount accessor>.name`                                  | Entity alias name for the given mount                                                   |
| `identity.entity.aliases.<mount accessor>.metadata`                              | Metadata associated with the alias for the given mount                                  |
| `identity.entity.aliases.<mount accessor>.metadata.<metadata key>`               | Metadata associated with the alias for the given mount and metadata key                 |
| `identity.entity.aliases.<mount accessor>.custom_metadata`                       | Custom metadata associated with the alias for the given mount                           |
| `identity.entity.aliases.<mount accessor>.custom_metadata.<custom_metadata key>` | Custom metadata associated with the alias for the given mount and custom metadata key   |
| `time.now`                                                                       | Current time as integral seconds since the Epoch                                        |
| `time.now.plus.<duration>`                                                       | Current time plus a Go-parsable [duration](https://golang.org/pkg/time/#ParseDuration)  |
| `time.now.minus.<duration>`                                                      | Current time minus a Go-parsable [duration](https://golang.org/pkg/time/#ParseDuration) |

```

> Several named scopes can be made available on an individual provider. Note that the top-level keys in a JSON template may conflict with those in another scope. When scopes are made available on a provider, their templates are checked for top-level conflicts. An error will occur if any conflict exists.

Reviewer: This needs to be clarified a bit. Suggestion below

Reviewer: ```suggestion
Several named scopes can be made available on an individual provider. Note that the top-level keys in a JSON template may conflict with those in another scope. When scopes are made available on a provider, their templates are checked for top-level conflicts. A warning will be issued to the vault operator if any conflicts are found. Note that this may result in an error if the scopes are requested in an OIDC Authentication Request.
```

> The openid scope is a unique case scope that may not be modified or deleted. The scope will exist in Vault and is supported by each provider by default. The scope represents the minimum set of claims required by the OIDC specification for inclusion in ID tokens. As such, templates may not contain top-level keys that overwrite the claims populated by the openid scope.

Reviewer: suggest `openid` to be in backticks

Author: Done

The following defines the claims key and value mapping for the openid scope:

* `iss`- configured issuer of the provider
* `sub`- unique entity ID of the Vault user

> * `aud`- iD of the client

Reviewer: ```suggestion
* `aud`- ID of the client
```

* `iat`- time of token issue
* `exp`- time of token issue + ID token TTL

### Client registration

> A client resource allows the relying party by providing metadata about itself to Vault to [dynamically register](https://openid.net/specs/openid-connect-registration-1_0.html).

Reviewer: ```suggestion
A client resource allows the relying party to [dynamically register](https://openid.net/specs/openid-connect-registration-1_0.html) by providing metadata about itself to Vault.
```

The client must have the following configuration parameters:

* **Redirect URIs**: limits the valid redirect URIs in an authentication request
* **Assignments**: determines who can authenticate with the client

> * **Key**: signing of the ID tokens

Reviewer: ```suggestion
* **Key**: used to sign the ID tokens
```

A `client_id` and `client_secret` are generated and returned after a successful client registration. Their values are strings using the base62 character set. The `client_id` will have 32 characters, and the `client_secret` will have a prefix of `hvo_secret`. The `client_id`  uniquely identifies the client. The `client_secret` will be used to authenticate to the token endpoint as described in [client authentication](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication).

The `key` parameter is required. The user must create a `key` as a required parameter of the client configuration.

~> **Note**: At least one of the redirect URIs of a client must exactly match the `redirect_uri` parameter used in an authentication request initiated by the client.

### Assignments

Assignment resources are referenced by clients via the `assignments` parameter. This parameter limits the set of Vault users allowed to authenticate. The assignments of an associated client are validated during the authentication request, ensuring that the Vault identity associated with the request is a member of the assignment's entities or groups.

### Keys

> Key resources are referenced by clients via the key parameter. This parameter specifies the key that will be used to sign ID tokens for the client. The key resource established is used to support identity tokens. See existing [documentation](/api-docs/secret/identity/tokens#create-a-named-key) for details on keyring management, supported signing algorithms, rotation periods, and verification TTLs. Currently, a key referenced by a client cannot be changed.

Reviewer: ```suggestion
Key resources are referenced by clients via the key parameter. This parameter specifies the key that will be used to sign ID tokens for the client. See existing [documentation](/api-docs/secret/identity/tokens#create-a-named-key) for details on keyring management, supported signing algorithms, rotation periods, and verification TTLs. Currently, a key referenced by a client cannot be changed.
```

Reviewer: I feel like maybe it is clearer without that sentence. Although maybe we do want to clarify that the `key` resource/api used by the Vault OIDC Provider feature is the same `key` resource/api used by the identity token system.

## OIDC flow

> The following sections provide implementation details for the OIDC compliant APIs provided by Vault OIDC providers.

Reviewer: I don't know if it makes sense to go here or at the top of this document. But we should clarify that the Vault OIDC Provider feature currently only supports the [authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth).

Below are some more details that should be mentioned in regards to the OIDC flow:

Vault OIDC providers will enable registered clients to authenticate and obtain identity information (or "claims") for their end-users. They will do so by providing the APIs and behavior required to satisfy the OIDC specification for the [authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth). Initially, all clients will be treated as first-party. This means that end-users will not be required to provide consent to the provider as detailed in [3.1.2.4](https://openid.net/specs/openid-connect-core-1_0.html#Consent) of the OIDC specification. The provider will release information to clients as long as the end-user has ACL access to the provider and their identity has been authorized via an assignment.

Author: Make sense. I added a note at the top of the document to specify that the feature is currently only supports the authorization code flow.

I also expanded the OIDC flow section to include your suggested text above. I've changed the tense a bit from passive to active in some places where I thought it might help. Sometimes, it's unavoidable and using passive voice might be necessary.

Reviewer: thanks!

### OpenID configuration

Each provider offers an unauthenticated endpoint that facilitates OIDC Discovery. All required metadata listed in [OpenID Provider Metadata](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata) is included in the discovery document. Additionally, the recommended `userinfo_endpoint` and `scopes_supported` metadata are included.

### Keys

> Each provider offers an unauthenticated endpoint that provides the public portion of keys used to sign ID tokens. The keys are published in a JSON Web Key Set [(JWKS)](https://datatracker.ietf.org/doc/html/rfc7517) format. The keyset for an individual provider contains the keys referenced by all clients via the `allowed_clients` configuration parameter. A `Cache-Control` header to set based on responses, allowing clients to refresh their keys upon rotation. The `max-age` of the header is set based on the earliest rotation time of any of the keys in the keyset.

Reviewer: ```suggestion
Each provider offers an unauthenticated endpoint that provides the public portion of keys used to sign ID tokens. The keys are published in a JSON Web Key Set [(JWKS)](https://datatracker.ietf.org/doc/html/rfc7517) format. The keyset for an individual provider contains the keys referenced by all clients via the `allowed_client_ids` configuration parameter. A `Cache-Control` header to set based on responses, allowing clients to refresh their keys upon rotation. The `max-age` of the header is set based on the earliest rotation time of any of the keys in the keyset.
```

The endpoint will [validate](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequestValidation) client requests ensuring all required parameters are present and valid. The `redirect_uri` of the request will be validated against the client's `redirect_uris`. The requesting Vault entity will be validated against the client's `assignments`. An appropriate [error code](https://openid.net/specs/openid-connect-core-1_0.html#AuthError) is returned for invalid requests.

An authorization code is generated with a successful validation of the request. The authorization code is single-use and cached with a lifetime of approximately 5 minutes, which mitigates the risk of leaks. A response including the original `state` presented by the client and `code` will be returned to the Vault UI and initiates the request. Vault will issue an HTTP 302 redirect to the `redirect_uri` of the request, which includes the `code`, `state`, and `nonce` as query parameters.

### Token Endpoint

> Each provider will offer a [token endpoint](https://openid.net/specs/openid-connect-core-1_0.html#TokenEndpoint). The endpoint may be unauthenticated in Vault but is authenticated by requiring a `client_secret` as described in [client authentication](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication). The endpoint ingests all required [token request](https://openid.net/specs/openid-connect-core-1_0.html#TokenRequest) parameters as input. The endpoint [validates](https://openid.net/specs/openid-connect-core-1_0.html#TokenRequestValidation) the client requests and exchanges an authorization code for the ID token and access token. The cache of authorization codes will be verified against the code presented in the exchange. The appropriate [error code](https://openid.net/specs/openid-connect-core-1_0.html#TokenErrorResponse) are returned for all invalid requests.

Reviewer: But I do think referencing the spec makes sense in some places

> Each provider will offer a [token endpoint](https://openid.net/specs/openid-connect-core-1_0.html#TokenEndpoint). The endpoint may be unauthenticated in Vault but is authenticated by requiring a `client_secret` as described in [client authentication](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication). The endpoint ingests all required [token request](https://openid.net/specs/openid-connect-core-1_0.html#TokenRequest) parameters as input. The endpoint [validates](https://openid.net/specs/openid-connect-core-1_0.html#TokenRequestValidation) the client requests and exchanges an authorization code for the ID token and access token. The cache of authorization codes will be verified against the code presented in the exchange. The appropriate [error code](https://openid.net/specs/openid-connect-core-1_0.html#TokenErrorResponse) are returned for all invalid requests.

Reviewer: `The appropriate [error code]` -> `The appropriate [error codes]`

> Each provider will offer a [token endpoint](https://openid.net/specs/openid-connect-core-1_0.html#TokenEndpoint). The endpoint may be unauthenticated in Vault but is authenticated by requiring a `client_secret` as described in [client authentication](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication). The endpoint ingests all required [token request](https://openid.net/specs/openid-connect-core-1_0.html#TokenRequest) parameters as input. The endpoint [validates](https://openid.net/specs/openid-connect-core-1_0.html#TokenRequestValidation) the client requests and exchanges an authorization code for the ID token and access token. The cache of authorization codes will be verified against the code presented in the exchange. The appropriate [error code](https://openid.net/specs/openid-connect-core-1_0.html#TokenErrorResponse) are returned for all invalid requests.

Author: I did replace some spec links with the api link, but for this one, I'll definitely need your help to cite the correct link(s) to replace the spec links, and determine where it makes sense to leave the spec links as is.

The ID token is generated and returned upon successful client authentication and request validation. The ID token will contain a combination of required and configurable claims. The required claims are enumerated in the scopes section above for the `openid` scope. The configurable claims are populated by templates associated with the scopes provided in the authentication request that generated the authorization code.

> An access token is also generated and returned upon successful client authentication and request validation. The access token is a Vault [batch token](https://www.vaultproject.io/docs/concepts/tokens#batch-tokens) with a policy that only provides read access to the issuing provider's [userinfo endpoint](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo). The access token is also a TTL as defined by the `access_token_ttl` of the requesting client.

Reviewer: Maybe link to vault API docs for the userinfo endpoint: https://github.com/hashicorp/vault/blame/main/website/content/api-docs/secret/identity/oidc-provider.mdx#L705

Author: added the preliminary links but they won't be hot until we release 1.9.

### UserInfo Endpoint

> Each provider provides an authenticated [userinfo endpoint](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo). The endpoint accepts the access token obtained from the token endpoint as a [bearer token](/api-docs#authentication). The userinfo response is a JSON object with the `application/json` content type. The JSON object contains claims for the Vault entity associated with the access token. The claims returned are determined by the scopes requested in the authentication request that produced the access token. The `sub` claim is always returned as the entity ID in the userinfo response.

Reviewer: Not sure if it makes sense to also link to vault API docs for the userinfo endpoint: https://github.com/hashicorp/vault/blame/main/website/content/api-docs/secret/identity/oidc-provider.mdx#L705

Or maybe both the OIDC spec and vault API docs?

Author: If you feel strongly about including both links, I can certainly add the spec link here as well, but it's nice to avoid including outside links if they can accomplish whatever they need to accomplish by accessing our api doc link. I'll defer that decision to you.

> ---

Reviewer 2: I think this document works pretty well as a conceptual explainer. However, I wonder if we should write an additional document that walks our users through the story of setting up OIDC provider with code snippets (similar to how our docs usually do this [example](https://www.vaultproject.io/docs/secrets/transform#setup)). 

Do we have plans to write additional documentation around OIDC provider? If not, I think we should adapt the structure of this page a bit and consider moving it out of the "Concepts" section. I would expect the document to live under "Secrets Engines" -> "Identity" -> "OIDC Provider". This matches up with the hierarchy of the API documentation for the feature. We could likely break out the identity token docs into its own section to distinguish the two use cases. Something like "Secrets Engines" -> "Identity" -> "Tokens".

Author: @austingebauer , hi. Yes, the education engineer (Lynn) will provide a learn tutorial on how to set up OIDC Provider. I can link to that tutorial from this document.

layout: docs
page_title: OIDC Provider
description: >-
  Describes how Vault can be an OIDC identity provider.
---

# OIDC Provider

> 

Reviewer 2: We should emphasize that this is a tech preview feature using something like:
```
~> **Note:** This feature is currently a ***Tech Preview*** and not recommended for deployment in production.
```

> ~> **Note**: The Vault OIDC Provider feature currently only supports the [authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth).

Reviewer 2: I'm not sure that this detail needs an emphasized callout at the top of the document. I suggest that we incorporate this into a "Supported Flows" section or note it in sections covering OIDC-specific concepts.

> This document describes how Vault can be an **OpenID Connect(OIDC) identity provider** by enabling applications to leverage Vault as a source of identity using the OIDC protocol.

Reviewer 2: ```suggestion
This document describes how Vault can be an **OpenID Connect (OIDC) identity provider** by enabling applications to leverage Vault as a source of identity using the OIDC protocol.
```

> This feature allows clients speaking to the OIDC protocol to take advantage of Vault's various authentication methods and source of identity. Clients can configure their authentication logic to talk to Vault. Once enabled, Vault will act as the bridge to identity providers via its existing authentication methods. Clients will also obtain identity information for their end-users by leveraging custom templating of Vault identity information.

Reviewer: ```suggestion
This feature allows clients speaking the OIDC protocol to take advantage of Vault's various authentication methods and source of identity. Clients can configure their authentication logic to talk to Vault. Once enabled, Vault will act as the bridge to identity providers via its existing authentication methods. Clients will also obtain identity information for their end-users by leveraging custom templating of Vault identity information.
```

Vault as an OIDC provider allows mutual Vault and Boundary customers to leverage Vault's identity system to delegate authentication and authorization to Vault.  Vault, therefore, acts as an identity provider for Boundary. Other HashiCorp products such as Consul can also leverage Vault's identity system and provide delegated authentication and authorization to its users. Having Vault as an OIDC provider allows a single sign-on experience to their end-users for organizations that want to leverage Vault as an identity provider.

## Configuration Options

The next few sections of the document provide implementation details for each resource that permits Vault configuration as an OIDC identity provider.

### Providers

A Vault namespace may contain several provider resources. Each configured provider will publish the APIs listed within the OIDC flow. The APIs will be served via backend path-based routing on Vault's listen [address](/docs/configuration/listener/tcp#address).

A provider must have the following configuration parameters:

* **Issuer URL**: used in the `iss` claim of ID tokens
* **Allowed client IDs**: limits which clients can access the provider

> * **Scopes supported**: limits what identity information is available as claims.

Reviewer 2: Nit: consistency with period at end of list items

Author: fixed.

### Scopes

Providers may reference scope resources via the `scopes_supported` parameter to make specific identity information available as claims.

A scope will have the following configuration parameters:

* **Description**: identity information captured by the scope
* **Template**: maps individual claims to Vault identity information

The template parameter takes advantage of the [JSON-based templating](/api-docs/secret/identity/tokens#template) used by identity tokens for claims mapping. This means the parameter will take a JSON string of arbitrary structure where the values may be replaced with specific identity information. Template parameters that are not present for a Vault identity are omitted from the resulting claims without an error.

Example of a JSON template for a scope:

```
{
  "groups": {{identity.entity.groups.names}},
  "email": {{identity.entity.metadata.user_email}},

>   "auth0": {

Reviewer 2: I wonder if we should use an example template similar to the one used for identity tokens [link](https://www.vaultproject.io/docs/secrets/identity#token-contents-and-templates). `auth0` as a claim was a bit confusing to me here being that it's an OIDC provider itself.

The `openid` scope is a unique case scope that may not be modified or deleted. The scope will exist in Vault and supported by each provider by default. The scope represents the minimum set of claims required by the OIDC specification for inclusion in ID tokens. As such, templates may not contain top-level keys that overwrite the claims populated by the openid scope.

The following defines the claims key and value mapping for the openid scope:

* `iss`- configured issuer of the provider
* `sub`- unique entity ID of the Vault user
* `aud`- ID of the client
* `iat`- time of token issue
* `exp`- time of token issue + ID token TTL

### Client registration

A client resource allows the relying party to [dynamically register](https://openid.net/specs/openid-connect-registration-1_0.html) by providing metadata about itself to Vault.

The client must have the following configuration parameters:

* **Redirect URIs**: limits the valid redirect URIs in an authentication request
* **Assignments**: determines who can authenticate with the client
* **Key**: used to sign the ID tokens

> *	**ID token TTL**: specifies the time-to-live for ID tokens

Reviewer 2: Indentation appears off here.

Author: fixed.

### Keys

Key resources are referenced by clients via the key parameter. This parameter specifies the key that will be used to sign ID tokens for the client. See existing [documentation](/api-docs/secret/identity/tokens#create-a-named-key) for details on keyring management, supported signing algorithms, rotation periods, and verification TTLs. Currently, a key referenced by a client cannot be changed.

## OIDC flow

The following sections provide implementation details for the OIDC compliant APIs provided by Vault OIDC providers.

> Vault OIDC providers enable registered clients to authenticate and obtain identity information (or "claims") for their end-users. They do this by providing the APIs and behavior required to satisfy the OIDC specification for the [authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth). Initially, all clients are treated as first-party. This means that end-users will not be required to provide consent to the provider as detailed in section [3.1.2.4](https://openid.net/specs/openid-connect-core-1_0.html#Consent) of the OIDC specification. The provider will release information to clients as long as the end-user has ACL access to the provider and their identity has been authorized via an assignment.

Reviewer: ```suggestion
Vault OIDC providers enable registered clients to authenticate and obtain identity information (or "claims") for their end-users. They do this by providing the APIs and behavior required to satisfy the OIDC specification for the [authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowAuth). All clients are treated as first-party. This means that end-users will not be required to provide consent to the provider as detailed in section [3.1.2.4](https://openid.net/specs/openid-connect-core-1_0.html#Consent) of the OIDC specification. The provider will release information to clients as long as the end-user has ACL access to the provider and their identity has been authorized via an assignment.
```

Reviewer: After reading this again, I think we should just state what we currently do. The "initially" is a little confusing.

### Keys

Each provider offers an unauthenticated endpoint that provides the public portion of keys used to sign ID tokens. The keys are published in a JSON Web Key Set [(JWKS)](https://datatracker.ietf.org/doc/html/rfc7517) format. The keyset for an individual provider contains the keys referenced by all clients via the `allowed_client_ids` configuration parameter. A `Cache-Control` header to set based on responses, allowing clients to refresh their keys upon rotation. The `max-age` of the header is set based on the earliest rotation time of any of the keys in the keyset.

### Authorization Endpoint

Each provider offers an authenticated [authorization endpoint](https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationEndpoint). The authorization endpoint for each provider is added to Vault's [default policy](/docs/concepts/policies#default-policy) using the `identity/oidc/provider/+/authorize` path. The endpoint incorporates all required [authentication request](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) parameters as input. Additionally, the `state` and `nonce` parameters are required.

> The endpoint [validates](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequestValidation) client requests and ensures that all required parameters are present and valid. The `redirect_uri` of the request are validated against the client's `redirect_uris`. The requesting Vault entity will be validated against the client's `assignments`. An appropriate [error code](https://openid.net/specs/openid-connect-core-1_0.html#AuthError) is returned for invalid requests.

Reviewer: ```suggestion
The endpoint [validates](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequestValidation) client requests and ensures that all required parameters are present and valid. The `redirect_uri` of the request is validated against the client's `redirect_uris`. The requesting Vault entity will be validated against the client's `assignments`. An appropriate [error code](https://openid.net/specs/openid-connect-core-1_0.html#AuthError) is returned for invalid requests.
```

An authorization code is generated with a successful validation of the request. The authorization code is single-use and cached with a lifetime of approximately 5 minutes, which mitigates the risk of leaks. A response including the original `state` presented by the client and `code` will be returned to the Vault UI which initiated the request. Vault will issue an HTTP 302 redirect to the `redirect_uri` of the request, which includes the `code`, `state`, and `nonce` as query parameters.

### Token Endpoint

> Each provider will offer a [token endpoint](/api-docs/secret/identity/oidc-provider.mdx). The endpoint may be unauthenticated in Vault but is authenticated by requiring a `client_secret` as described in [client authentication](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication). The endpoint ingests all required [token request](/api-docs/secret/identity/oidc-provider.mdx) parameters as input. The endpoint [validates](/api-docs/secret/identity/oidc-provider.mdx) the client requests and exchanges an authorization code for the ID token and access token. The cache of authorization codes will be verified against the code presented in the exchange. The appropriate [error codes](https://openid.net/specs/openid-connect-core-1_0.html#TokenErrorResponse) are returned for all invalid requests.

Reviewer: ```suggestion
Each provider will offer a [token endpoint](/api-docs/secret/identity/oidc-provider#token-endpoint). The endpoint may be unauthenticated in Vault but is authenticated by requiring a `client_secret` as described in [client authentication](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication). The endpoint ingests all required [token request](/api-docs/secret/identity/oidc-provider#parameters-15) parameters as input. The endpoint [validates](https://openid.net/specs/openid-connect-core-1_0.html#TokenRequestValidation) the client requests and exchanges an authorization code for the ID token and access token. The cache of authorization codes will be verified against the code presented in the exchange. The appropriate [error codes](https://openid.net/specs/openid-connect-core-1_0.html#TokenErrorResponse) are returned for all invalid requests.
```

Reviewer: I think those are the correct anchors. Although I am not sure if we usually link to parameters like that. Maybe we could just link to the `#token-endpoint` if not.

I think for the validation it makes sense to link to the OIDC spec.
